### PR TITLE
HPCC-25085 Exclude Indexes from file list

### PR DESCRIPTION
**Pull request by kenrowland**
Added option to invert search results to retrieve only
non index file list

Signed-off-by: Ken Rowland kenneth.rowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by wangkx — APPROVED** [review]
Looks good to me.

**Comment by kenrowland** [comment]
@richardkchapman Please merge

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 1]
```diff
@@ -3612,7 +3612,15 @@ void CWsDfuEx::setDFUQueryFilters(IEspDFUQueryRequest& req, StringBuffer& filter
     setFileNameFilter(req.getLogicalName(), req.getPrefix(), filterBuf);
     setFileTypeFilter(req.getFileType(), filterBuf);
     WsDFUHelpers::appendDFUQueryFilter(getDFUQFilterFieldName(DFUQFFattrowner), DFUQFTwildcardMatch, req.getOwner(), filterBuf);
-    WsDFUHelpers::appendDFUQueryFilter(getDFUQFilterFieldName(DFUQFFkind), DFUQFTwildcardMatch, req.getContentType(), filterBuf);
+
+    if (req.getInvertContent_isNull() || !req.getInvertContent())
+    {
+        WsDFUHelpers::appendDFUQueryFilter(getDFUQFilterFieldName(DFUQFFkind), DFUQFTwildcardMatch, req.getContentType(), filterBuf);
```
Comment (jakesmith) [inline esp/services/ws_dfu/ws_dfuService.cpp:3618]: question/not new: if ContentType is not set - why is building/adding it as a filter ?

**Comment by jakesmith** [comment]
Not convinced this feature should be implemented this way - i.e. by request having to set ContentType="key" and InvertContent=true.

(from JIRA description) - "Currently the filter for a file list allows you to include only indexes, but I can't see a way to exclude indexes."

I would have expected this to be implemented by being able to select the file type via a drop-down and therefore choose e.g. "flat" files only. NB: it may be that some files are of unknown type if sprayed.

@ghalliday - may want to comment on this.

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 1 above)
Comment (kenrowland) [reply esp/services/ws_dfu/ws_dfuService.cpp:3618]: @jakesmith the current logic was not changed except for honoring the invert content flag. The appendDFUQueryFilter method simply returns if there is no filter to add.  IMO if that logic should be changed, a separate Jira should be opened.

Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp:3618]: >The appendDFUQueryFilter method simply returns if there is no filter to add. 

ok, that's fine.

**Comment by jakesmith** [comment]
> The UI sets the ContentType. It can expand to use a dropdown (probably a separate Jira). The new InvertContent flag returns files that don't match the ContentType value. So, to get non-index files, set ContentType to 'key' and the InvertContent flag to true.
> 
> Think of setting the InvertContent flag as making the compare a !=.


I understood the logic/implementation. I was questioning whether it was the right implementation.
I think it is a bit overcomplicated, and only allows the inverse of 1 specific field.

There may be something I'm missing, but why does the DFUQueryRequest need extending?
i.e. why isn't ContentType sufficient already? 
NB: I know the UI only has an Index checkbox, but the ecm request definition looks correct,
it's the UI that is limiting functionality as far as I can see.

**Comment by kenrowland** [comment]
@ghalliday - Jake asked if you were interested in commenting.

**Review by jakesmith — APPROVED** [review]
@kenrowland - approved. A rather specific approach (inverse of 1 field type only), rather than a general mechanism, or a way to specifically select the types you want to see, but it provides the functionality requested, can be revisited at another time if we want to generalize/improve.

**Comment by kenrowland** [comment]
@richardkchapman Please merge (not critical)

**Pull request merged**